File: eprwebui/src/services/materialCompositionService.ts
import { MaterialComposition, CreateMaterialCompositionRequest } from '../types/materialComposition'

const API_BASE_URL = 'http://localhost:8080/api'

interface RequestOptions {
  token?: string
}

class MaterialCompositionService {
  private async makeRequest<T>(
    endpoint: string, 
    options: RequestInit & RequestOptions = {}
  ): Promise<T> {
    const { token, ...fetchOptions } = options
    
    const headers: Record<string, string> = {
      'Content-Type': 'application/json',
      ...fetchOptions.headers as Record<string, string>
    }

    // TEMPORARY: Skip authorization header for testing
    // if (token) {
    //   headers['Authorization'] = `Bearer ${token}`
    // }

    const response = await fetch(`${API_BASE_URL}${endpoint}`, {
      ...fetchOptions,
      headers
    })

    if (!response.ok) {
      const errorText = await response.text()
      throw new Error(errorText || `HTTP error! status: ${response.status}`)
    }

    // Handle empty responses (like DELETE operations)
    if (response.status === 204 || response.headers.get('content-length') === '0') {
      return {} as T
    }

    return response.json()
  }

  // Get all material compositions
  async getAllCompositions(token?: string): Promise<MaterialComposition[]> {
    return this.makeRequest<MaterialComposition[]>('/material-compositions', { token })
  }

  // Get material composition by ID
  async getCompositionById(id: number, token?: string): Promise<MaterialComposition> {
    return this.makeRequest<MaterialComposition>(`/material-compositions/${id}`, { token })
  }

  // Get compositions by material ID
  async getCompositionsByMaterialId(materialId: number, token?: string): Promise<MaterialComposition[]> {
    return this.makeRequest<MaterialComposition[]>(`/material-compositions/material/${materialId}`, { token })
  }

  // Search material compositions
  async searchCompositions(query: string, token?: string): Promise<MaterialComposition[]> {
    return this.makeRequest<MaterialComposition[]>(`/material-compositions/search?query=${encodeURIComponent(query)}`, { token })
  }

  // Get compositions by active status
  async getCompositionsByActiveStatus(isActive: boolean, token?: string): Promise<MaterialComposition[]> {
    return this.makeRequest<MaterialComposition[]>(`/material-compositions/active/${isActive}`, { token })
  }

  // Create a new material composition
  async createComposition(composition: CreateMaterialCompositionRequest, token?: string): Promise<MaterialComposition> {
    return this.makeRequest<MaterialComposition>('/material-compositions', {
      method: 'POST',
      body: JSON.stringify(composition),
      token
    })
  }

  // Update an existing material composition
  async updateComposition(id: number, composition: Partial<MaterialComposition>, token?: string): Promise<MaterialComposition> {
    return this.makeRequest<MaterialComposition>(`/material-compositions/${id}`, {
      method: 'PUT',
      body: JSON.stringify(composition),
      token
    })
  }

  // Delete a material composition
  async deleteComposition(id: number, token?: string): Promise<void> {
    return this.makeRequest<void>(`/material-compositions/${id}`, {
      method: 'DELETE',
      token
    })
  }

  // Get count of active compositions
  async getActiveCompositionsCount(token?: string): Promise<number> {
    return this.makeRequest<number>('/material-compositions/count', { token })
  }

  // Get count of compositions by material
  async getCompositionsCountByMaterial(materialId: number, token?: string): Promise<number> {
    return this.makeRequest<number>(`/material-compositions/count/material/${materialId}`, { token })
  }

  // Find compositions by weight range
  async findCompositionsByWeightRange(minWeight: number, maxWeight: number, token?: string): Promise<MaterialComposition[]> {
    return this.makeRequest<MaterialComposition[]>(`/material-compositions/weight-range?minWeight=${minWeight}&maxWeight=${maxWeight}`, { token })
  }

  // Find compositions by percentage overlap
  async findCompositionsByPercentageOverlap(minPercentage: number, maxPercentage: number, token?: string): Promise<MaterialComposition[]> {
    return this.makeRequest<MaterialComposition[]>(`/material-compositions/percentage-overlap?minPercentage=${minPercentage}&maxPercentage=${maxPercentage}`, { token })
  }

  // Validation helpers
  validateComposition(composition: CreateMaterialCompositionRequest | Partial<MaterialComposition>): string[] {
    const errors: string[] = []

    if (!composition.compositionName?.trim()) {
      errors.push('Composition name is required')
    }

    if (!composition.compositionCode?.trim()) {
      errors.push('Composition code is required')
    }

    if (!composition.materialId) {
      errors.push('Material selection is required')
    }

    if (!composition.weightKg || composition.weightKg <= 0) {
      errors.push('Weight must be greater than 0')
    }

    if (composition.minPercentage === undefined || composition.minPercentage === null) {
      errors.push('Minimum percentage is required')
    } else if (composition.minPercentage < 0 || composition.minPercentage > 100) {
      errors.push('Minimum percentage must be between 0 and 100')
    }

    if (composition.maxPercentage === undefined || composition.maxPercentage === null) {
      errors.push('Maximum percentage is required')
    } else if (composition.maxPercentage < 0 || composition.maxPercentage > 100) {
      errors.push('Maximum percentage must be between 0 and 100')
    }

    if (composition.minPercentage !== undefined && composition.maxPercentage !== undefined) {
      if (composition.minPercentage > composition.maxPercentage) {
        errors.push('Minimum percentage cannot be greater than maximum percentage')
      }
    }

    return errors
  }

  // Format helpers
  formatWeight(weight: number): string {
    return `${weight.toFixed(3)} kg`
  }

  formatPercentageRange(minPercentage: number, maxPercentage: number): string {
    return `${minPercentage.toFixed(2)}% - ${maxPercentage.toFixed(2)}%`
  }

  formatPercentage(percentage: number): string {
    return `${percentage.toFixed(2)}%`
  }
}

export default new MaterialCompositionService()

File: eprapiservices/src/main/java/epr/eprapiservices/service/MaterialCompositionService.java
package epr.eprapiservices.service;

import epr.eprapiservices.entity.MaterialComposition;
import epr.eprapiservices.entity.Material;
import epr.eprapiservices.dao.repository.MaterialCompositionRepository;
import epr.eprapiservices.dao.repository.MaterialRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;
import java.util.List;
import java.util.Optional;

/**
 * Service class for MaterialComposition business logic
 */
@Service
@Transactional
public class MaterialCompositionService {

    @Autowired
    private MaterialCompositionRepository materialCompositionRepository;

    @Autowired
    private MaterialRepository materialRepository;

    /**
     * Get all active material compositions
     */
    @Transactional(readOnly = true)
    public List<MaterialComposition> getAllActiveCompositions() {
        return materialCompositionRepository.findAllActiveOrderBySortOrderAndName();
    }

    /**
     * Get material composition by ID
     */
    @Transactional(readOnly = true)
    public Optional<MaterialComposition> getCompositionById(Integer id) {
        return materialCompositionRepository.findByIdAndActive(id);
    }

    /**
     * Get compositions by material ID
     */
    @Transactional(readOnly = true)
    public List<MaterialComposition> getCompositionsByMaterialId(Integer materialId) {
        return materialCompositionRepository.findByMaterialIdAndActive(materialId);
    }

    /**
     * Search compositions by term
     */
    @Transactional(readOnly = true)
    public List<MaterialComposition> searchCompositions(String searchTerm) {
        if (searchTerm == null || searchTerm.trim().isEmpty()) {
            return getAllActiveCompositions();
        }
        return materialCompositionRepository.searchByNameCodeOrDescription(searchTerm.trim());
    }

    /**
     * Get compositions by active status
     */
    @Transactional(readOnly = true)
    public List<MaterialComposition> getCompositionsByActiveStatus(Boolean isActive) {
        return materialCompositionRepository.findByActiveStatus(isActive);
    }

    /**
     * Create a new material composition
     */
    public MaterialComposition createComposition(MaterialComposition composition) {
        // Validate required fields
        validateComposition(composition, null);

        // Validate material exists
        validateMaterialExists(composition.getMaterialId());

        // Check for duplicate composition code
        if (materialCompositionRepository.existsByCompositionCodeAndActiveExcludingId(
                composition.getCompositionCode(), null)) {
            throw new RuntimeException("Composition code already exists: " + composition.getCompositionCode());
        }

        // Validate percentage range
        validatePercentageRange(composition.getMinPercentage(), composition.getMaxPercentage());

        // Set default values
        if (composition.getSortOrder() == null) {
            composition.setSortOrder(1);
        }
        composition.setIsActive(true);

        return materialCompositionRepository.save(composition);
    }

    /**
     * Update an existing material composition
     */
    public MaterialComposition updateComposition(Integer id, MaterialComposition compositionDetails) {
        Optional<MaterialComposition> existingCompositionOpt = materialCompositionRepository.findByIdAndActive(id);
        
        if (existingCompositionOpt.isPresent()) {
            MaterialComposition existingComposition = existingCompositionOpt.get();
            
            // Validate updated composition
            validateComposition(compositionDetails, id);

            // Validate material exists
            validateMaterialExists(compositionDetails.getMaterialId());

            // Check for duplicate composition code (excluding current record)
            if (materialCompositionRepository.existsByCompositionCodeAndActiveExcludingId(
                    compositionDetails.getCompositionCode(), id)) {
                throw new RuntimeException("Composition code already exists: " + compositionDetails.getCompositionCode());
            }

            // Validate percentage range
            validatePercentageRange(compositionDetails.getMinPercentage(), compositionDetails.getMaxPercentage());

            // Update fields
            existingComposition.setCompositionName(compositionDetails.getCompositionName());
            existingComposition.setCompositionCode(compositionDetails.getCompositionCode());
            existingComposition.setDescription(compositionDetails.getDescription());
            existingComposition.setMaterialId(compositionDetails.getMaterialId());
            existingComposition.setWeightKg(compositionDetails.getWeightKg());
            existingComposition.setMinPercentage(compositionDetails.getMinPercentage());
            existingComposition.setMaxPercentage(compositionDetails.getMaxPercentage());
            existingComposition.setSortOrder(compositionDetails.getSortOrder());
            existingComposition.setNotes(compositionDetails.getNotes());

            return materialCompositionRepository.save(existingComposition);
        } else {
            throw new RuntimeException("Material composition not found with id: " + id);
        }
    }

    /**
     * Delete a material composition (soft delete)
     */
    public void deleteComposition(Integer id) {
        Optional<MaterialComposition> compositionOpt = materialCompositionRepository.findByIdAndActive(id);
        
        if (compositionOpt.isPresent()) {
            MaterialComposition composition = compositionOpt.get();
            composition.setIsActive(false);
            materialCompositionRepository.save(composition);
        } else {
            throw new RuntimeException("Material composition not found with id: " + id);
        }
    }

    /**
     * Get count of active compositions
     */
    @Transactional(readOnly = true)
    public long getActiveCompositionsCount() {
        return materialCompositionRepository.countActiveCompositions();
    }

    /**
     * Get count of compositions by material
     */
    @Transactional(readOnly = true)
    public long getCompositionsCountByMaterial(Integer materialId) {
        return materialCompositionRepository.countByMaterialIdAndActive(materialId);
    }

    /**
     * Find compositions by weight range
     */
    @Transactional(readOnly = true)
    public List<MaterialComposition> findCompositionsByWeightRange(BigDecimal minWeight, BigDecimal maxWeight) {
        return materialCompositionRepository.findByWeightRange(minWeight, maxWeight);
    }

    /**
     * Find compositions by percentage overlap
     */
    @Transactional(readOnly = true)
    public List<MaterialComposition> findCompositionsByPercentageOverlap(BigDecimal minPercentage, BigDecimal maxPercentage) {
        return materialCompositionRepository.findByPercentageOverlap(minPercentage, maxPercentage);
    }

    // Private validation methods
    private void validateComposition(MaterialComposition composition, Integer excludeId) {
        if (composition.getCompositionName() == null || composition.getCompositionName().trim().isEmpty()) {
            throw new RuntimeException("Composition name is required");
        }
        if (composition.getCompositionCode() == null || composition.getCompositionCode().trim().isEmpty()) {
            throw new RuntimeException("Composition code is required");
        }
        if (composition.getMaterialId() == null) {
            throw new RuntimeException("Material ID is required");
        }
        if (composition.getWeightKg() == null || composition.getWeightKg().compareTo(BigDecimal.ZERO) <= 0) {
            throw new RuntimeException("Weight must be greater than 0");
        }
        if (composition.getMinPercentage() == null) {
            throw new RuntimeException("Minimum percentage is required");
        }
        if (composition.getMaxPercentage() == null) {
            throw new RuntimeException("Maximum percentage is required");
        }
    }

    private void validateMaterialExists(Integer materialId) {
        Optional<Material> material = materialRepository.findById(materialId);
        if (material.isEmpty() || !material.get().getIsActive()) {
            throw new RuntimeException("Material not found or inactive with id: " + materialId);
        }
    }

    private void validatePercentageRange(BigDecimal minPercentage, BigDecimal maxPercentage) {
        if (minPercentage.compareTo(BigDecimal.ZERO) < 0 || minPercentage.compareTo(new BigDecimal("100")) > 0) {
            throw new RuntimeException("Minimum percentage must be between 0 and 100");
        }
        if (maxPercentage.compareTo(BigDecimal.ZERO) < 0 || maxPercentage.compareTo(new BigDecimal("100")) > 0) {
            throw new RuntimeException("Maximum percentage must be between 0 and 100");
        }
        if (minPercentage.compareTo(maxPercentage) > 0) {
            throw new RuntimeException("Minimum percentage cannot be greater than maximum percentage");
        }
    }
}
